ark/spLDA/processing.py

```diff
@@ -63,24 +52,153 @@ def format_cell_table(cell_table, markers=None, clusters=None, fovs="all_fovs"):
+                                                        "markers"])
+    else:
+        neighborhood_feature_fn = functools.partial(func_type[feature_by])
+
+    # Featurize FOVs
+    feature_sample = {k: v for (k, v) in cell_table.items() if
+                      k in cell_table["fovs"].tolist()}
+    featurized_fovs = featurize_samples(feature_sample,
+                                        neighborhood_feature_fn,
+                                        radius=radius,
+                                        is_anchor_col=cell_index,
+                                        x_col='x',
+                                        y_col='y',
+                                        n_processes=n_processes,
+                                        include_anchors=True)
+
+    if train_split:
+        all_sample_idxs = featurized_fovs.index.map(lambda x: x[0])
+        _sets = train_test_split(featurized_fovs, test_size=1. - train_frac,
+                                 stratify=all_sample_idxs)
+        train_features_fraction, _ = _sets
+    else:
+        train_features_fraction = None
+
+    return {"featurized_fovs": featurized_fovs,
+            "train_features": train_features_fraction}
+
+
+def create_difference_matrices(cell_table, features, training=True,
+                               inference=True):
+    """Constructs the difference matrices used for training and inference for
+    each field of view in the formatted cell table.
 
+    Args:
+        cell_table (dict, pd.DataFrame):
+            A formatted cell table for use in spatial-LDA analysis.
+            Specifically, this is the output from
+            format_cell_table().
+        features (dict, pd.DataFrame):
+            A featurized cell table and training split. Specifically, this is
+            the output from featurize_cell_table().
+        training (bool):
+            If True, create the difference matrix for running training
+            algorithm.  One or both of training and inference
+            must be True.
+        inference (bool):
+             If True, create the difference matrix for running inference
+             algorithm.
 
+    Returns:
+        dict:
 
+        - A dictionary containing the training and inference difference
+        matrices.
+    """
+    cell_table = {k: v for (k, v) in cell_table.items() if
+                  k not in ["fovs", "markers", "clusters"]}
+    # check args function here
+    if training:
+        train_diff_mat = make_merged_difference_matrices(
+            sample_features=features["train_features"],
+            sample_dfs=cell_table,
+            x_col="x", y_col="y")
+    else:
+        train_diff_mat = None
+
+    if inference:
+        inference_diff_mat = make_merged_difference_matrices(
+            sample_features=features["featurized_fovs"],
+            sample_dfs=cell_table,
+            x_col="x", y_col="y")
+    else:
+        inference_diff_mat = None
+
```

Comment:
I don't understand the point of this function. How could you be doing training and inference at the same time? Why do we need this wrapper around make_merged_difference_matrices?

Reply:
Both the train() and infer() functions from the spatial_lda library take a difference matrix as one of their arguments.  In train(), the difference matrix is computed using the training fraction from the featurized data and the full formatted data.  In infer(), the difference matrix is calculated using the full featurized data and the full formatted data.  The wrapper function computes one of (or both of) these in one step so they can be passed to the respective training and inference functions later rather than having the user explicitly call them separately -- less steps for them.

Reply:
Okay, maybe once I see how this gets called it will make more sense to me. Right now, it seems like you would call this function once during training, and once during inference. But calling this function isn't any less work than calling make_merged_difference_matrices, so I'm not sure I see the benefit.